### PR TITLE
TEPHRA-228 Adding the ability to pass-in a clientId during the start …

### PR DESCRIPTION
**Pull request by gokulavasan**
…of a transaction which is logged when the transaction gets invalidated during time out.

### CODE REVIEW

File: tephra-core/src/main/java/org/apache/tephra/TransactionManager.java [hunk 1]
```diff
@@ -763,15 +781,22 @@ private long getNextWritePointer() {
    * transaction moves it to the invalid list because we assume that its writes cannot be rolled back.
    */
   public Transaction startLong() {
+    return startLong(null);
+  }
+
+  /**
+   * Starts a long transaction with a client id.
+   */
+  public Transaction startLong(String clientId) {
```
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: If we are going to deprecate the clientId-less methods, should we allow null here?

File: tephra-core/src/main/java/org/apache/tephra/TransactionManager.java [hunk 2]
```diff
@@ -1378,17 +1403,32 @@ public TransactionType getTransactionType() {
     private final long expiration;
     private final InProgressType type;
     private final LongArrayList checkpointWritePointers;
+    private final String clientId;
```
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: Not sure if we can add it to equals/hashCode. The unit tests fail when trying to compare the state before persistence and after recovery and won't be the same if we include clientId in comparison. Any reason why we should definitely include it? Or should we include it and adjust the test that does the comparison?

File: tephra-core/src/main/java/org/apache/tephra/TransactionManager.java (hunk 1 above)
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: Should I add Nullable annotation if we decide to allow null for clientId?

Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: Eventually we want all client's to pass in a client i, hence the deprecation. If we allow nulls then that defeats the purpose, no?

Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: @poornachandra Sure sounds good. I will add a Precondition check for String not being null.

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments. Please take a look.

File: tephra-core/src/main/java/org/apache/tephra/TransactionSystemClient.java [hunk 3]
```diff
@@ -44,11 +44,33 @@
   Transaction startShort(int timeout);
```
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TransactionSystemClient.java]: Done.

File: tephra-core/src/main/java/org/apache/tephra/TransactionManager.java (hunk 1 above)
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: Added Preconditions null check for clientId.

File: tephra-core/src/main/java/org/apache/tephra/TransactionManager.java (hunk 2 above)
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/TransactionManager.java]: It is okay not to add it now since we are not persisting it. Add a comment in the code.

File: tephra-hbase-compat-0.98/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java [hunk 4]
```diff
@@ -127,7 +126,7 @@ public TransactionProcessor() {
   public void start(CoprocessorEnvironment e) throws IOException {
     if (e instanceof RegionCoprocessorEnvironment) {
       RegionCoprocessorEnvironment env = (RegionCoprocessorEnvironment) e;
-      Supplier<TransactionStateCache> cacheSupplier = getTransactionStateCacheSupplier(env);
+      this.cacheSupplier = getTransactionStateCacheSupplier(env);
```
Comment (poornachandra) [inline tephra-hbase-compat-0.98/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java]: Unrelated change?

Comment (gokulavasan) [inline tephra-hbase-compat-0.98/src/main/java/org/apache/tephra/hbase/coprocessor/TransactionProcessor.java]: Missed this change in https://github.com/apache/incubator-tephra/pull/41 for this compat-module. Hence fixing it here.

**Comment by gokulavasan** [comment]
@poornachandra Can you please review the last commit. I have addressed your comment about TransactionContext change (deprecated old constructors and introduced a new one with the clientId)

**Comment by gokulavasan** [comment]
@poornachandra Injecting ClientId in TransactionServiceClient now instead of changing the API of TransactionSystemClient. Please review when you get a chance.

**Comment by gokulavasan** [comment]
@poornachandra Please review the last commit. I have addressed your comment about including hostname-pid as the default client id.

File: tephra-core/src/main/java/org/apache/tephra/runtime/ClientIdAwareTransactionModule.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.google.inject.AbstractModule;
+import com.google.inject.name.Names;
+import org.apache.tephra.TransactionSystemClient;
+import org.apache.tephra.distributed.TransactionServiceClient;
+
+/**
+ * Guice module that binds the clientId to be used in {@link TransactionSystemClient}.
+ */
+public abstract class ClientIdAwareTransactionModule extends AbstractModule {
```
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/runtime/ClientIdAwareTransactionModule.java]: Having `ClientIdAwareTransactionModule` as the parent class of Transaction modules does not add much value. Let's move the binding to the child modules.

File: tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java [hunk 6]
```diff
@@ -35,13 +34,24 @@
...
+import java.lang.management.ManagementFactory;
+
 /**
  * Guice bindings for running in distributed mode on a cluster.
  */
-final class TransactionDistributedModule extends AbstractModule {
+final class TransactionDistributedModule extends ClientIdAwareTransactionModule {
+
+  public TransactionDistributedModule() {
+    super(ManagementFactory.getRuntimeMXBean().getName());
```
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java]: This default is already defined in `TransactionModules`. Do we still need it here? 
Same for `TransactionInMemoryModule` and `TransactionLocalModule`.

Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java]: @poornachandra There is no guarantee that TransactionModules().getDistributedModule() is used to get an instance of TransactionDistributedModule. One can just create an instance of this class.

File: tephra-core/src/main/java/org/apache/tephra/runtime/ClientIdAwareTransactionModule.java (hunk 5 above)
Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/runtime/ClientIdAwareTransactionModule.java]: Sure ok.

File: tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java (hunk 6 above)
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java]: In that case they can provide the client id since the default constructor won't be present, right?

**Comment by gokulavasan** [comment]
@poornachandra Thanks for the comments. Addressed them.

File: tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java [hunk 7]
```diff
@@ -130,10 +134,11 @@ public static void doMain(boolean verbose, Configuration conf) throws Exception
    * for service discovery. Otherwise it will look for the port in the
    * config and use localhost.
    * @param config a configuration containing the zookeeper properties
+   * @param clientId id of the client that identifies it when it starts a transaction
    */
   @Inject
-  public TransactionServiceClient(Configuration config,
-                                  ThriftClientProvider clientProvider) {
+  public TransactionServiceClient(Configuration config, ThriftClientProvider clientProvider,
```
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java]: Some clients like Apache Phoenix do not use the guice modules. They instantiate `TransactionServiceClient` directly. It will be good to have the existing constructor too and assign the default value for the clientId in it.

Comment (gokulavasan) [inline tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java]: @poornachandra Addressed comment in the commit https://github.com/apache/incubator-tephra/pull/42/commits/0279658b3b0718298ccfe363343bce182e4ba655. Please take a look again.

File: tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java [hunk 8]
```diff
@@ -125,15 +130,20 @@ public static void doMain(boolean verbose, Configuration conf) throws Exception
     }
   }
 
+  public TransactionServiceClient(Configuration config, ThriftClientProvider clientProvider) {
```
Comment (poornachandra) [inline tephra-core/src/main/java/org/apache/tephra/distributed/TransactionServiceClient.java]: Please add javadocs for the public constructor.
